src/main/java/org/reaktivity/reaktor/internal/budget/DefaultBudgetCreditor.java

```diff
@@ -143,17 +163,65 @@ public void creditById(
         long credit)
     {
         final long budgetIndex = budgetIndexById.get(budgetId);
+
+        if (ReaktorConfiguration.DEBUG_BUDGETS)
+        {
+            System.out.format("[%d] creditById credit=%d budgetId=%d budgetIndex=%d %s \n",
+                System.nanoTime(), credit, budgetId, budgetIndex, budgetIndexById.toString());
+        }
+
         if (budgetIndex != NO_CREDITOR_INDEX)
         {
             credit(traceId, budgetIndex, credit);
         }
     }
 
+    @Override
+    public long supplyChild(
+        long parentBudgetId)
+    {
+        final long childBudgetId = supplyBudgetId.getAsLong();
+        parentBudgetIds.put(childBudgetId, parentBudgetId);
+
+        return childBudgetId;
+    }
+
     public int acquired()
     {
         return budgetIndexById.size();
     }
 
+    public long parentBudget()
+    {
+        return parentBudgetIds.size();
+    }
+
+    public long parentBudgetId(
+        long budgetId)
+    {
+        long parentBudgetId = parentBudgetIds.get(budgetId);
+
+        if (parentBudgetId == NO_CREDITOR_INDEX)
+        {
+            parentBudgetId = budgetId;
+        }
+        return parentBudgetId;
+    }
+
+    @Override
+    public void cleanupChild(
+        long budgetId)
+    {
+        if (ReaktorConfiguration.DEBUG_BUDGETS)
+        {
+            System.out.format("[%d] cleanupChild childBudgetId=%d budgetParentChildRelation=%s \n",
+                System.nanoTime(), budgetId, parentBudgetIds.toString());
+        }
+        signaler.executeTaskAt(currentTimeMillis() + childCleanupLinger,
```

Comment:
Pass `Signaler` implementation instead to be able to use internal implementation, rather than introducing new public API.